File: sem3/map/Exam/GUI_ToyLanguageInterpreter/src/Interpreter/Models/Statements/IfStatement.java
package Interpreter.Models.Statements;

import Interpreter.Models.*;
import Interpreter.Models.ProgramState.ProgramState;
import Interpreter.Utils.MyIStack;

public class IfStatement implements IStatement {

    private Expression expression;
    private IStatement thenStatement;
    private IStatement elseStatement;

    public IfStatement(Expression expression, IStatement thenStatement, IStatement elseStatement) {

        this.expression = expression;
        this.thenStatement = thenStatement;
        this.elseStatement = elseStatement;
    }

    @Override
    public ProgramState execute(ProgramState state) throws Exception {
        MyIStack<IStatement> stack = state.getExeStack();

        int expressionValue = this.expression.eval(state);

        if (expressionValue != 0) {
            stack.push(this.thenStatement);
        } else {
            stack.push(this.elseStatement);
        }

        return null;
    }

    @Override
    public String toString() {

        return "if (" + this.expression.toString() + ") {\n" +
                this.thenStatement.toString() +
                "\nelse {\n" +
                this.elseStatement.toString() +
                "\n}\n";
    }
}

File: sem3/map/Lab/toylanguage_C#/ToyLanguage/ToyLanguage/Models/Statements/IfStatement.cs
using ToyLanguage.Models.Interfaces;
using ToyLanguage.Models.States;

namespace ToyLanguage.Models.Statements
{
    public class IfStatement : IStatement
    {
        private readonly IStatement _ifStatement;
        private readonly IExpression _expression;
        private readonly IStatement _elseStatement;

        public IfStatement(IStatement ifStatement, IExpression expression, IStatement elseStatement)
        {
            _ifStatement = ifStatement;
            _expression = expression;
            _elseStatement = elseStatement;
        }


        public ProgramState Execute(ProgramState state)
        {
            var expressionValue = _expression.Eval(state);

            if (expressionValue != 0)
            {
                state.ExecutionStack.Push(_ifStatement);
            }
            else
            {
                state.ExecutionStack.Push(_elseStatement);
            }

            return state;
        }

        public override string ToString()
        {
            return $"if ({_expression}) {{\n{_ifStatement}\n}} else {{\n{_elseStatement}\n}}";
        }
    }
}